extraConfigBFB.py

```diff
@@ -34,18 +34,18 @@ def ExtraConfigBFB(cc: ClustersConfig, _: ExtraConfigArgs, futures: Dict[str, Fu
     nfs = NFS(lh, cc.external_port)
     iso_url = nfs.host_file("/root/iso/fedora-coreos.iso")
 
-    def helper(h: host.HostWithBF2) -> Optional[host.Result]:
+    def helper(h: host.HostWithBF2, bmc: host.BMC) -> Optional[host.Result]:
```

Comment:
You're saying "host has nothing to do with BMC" and then you fix it by now having to pass `host` and `bmc` as two separate things around?

Aren't they tightly coupled if you need to path both around together?

Reply:
The are related, as a Host and BMC instance both are relevant for a particular host. Other than that, there is no logic or code that depends on each other.

Host itself can be quite simple. BMC is also simple. Mixing them together, makes them more complicated. E.g. the interning in `host.host_instances` needs to consider the BMC URL. On the other hand, passing both a `host` and `bmc` instance around, is something that you can hardly mess up, because you will notice when you need it but don't have it.

Note that the branch only changes few places, where now an additional BMC instance needs to be passed around. If this would have to happen frequently, then there would be more reason to attach the BMC to the host instance. Note that the patch does not cause an increase of lines of code.

See the first patch, which fixes a bug which probably wouldn't have been possible, if Host wouldn't pretend to also be a BMC.

Reply:
Does it make sense to have an instance of BMC without a Host or is the following statement true: "if I have a bmc, I also have a host associated with that bmc somewhere in the code".

Reply:
> "if I have a bmc, I also have a host associated with that bmc somewhere in the code"

note that this statement would sound as if a BMC should always hold a reference to a Host. Instead, we have the Host sometimes holding a BMC. It makes sense this way, but this "sometimes" makes it ugly (see those `raise Exception(f"Can't start host without bmc on {self.hostname()}")`).

See again commit `"dc9e029a78ff clusterSnapshotter: fix boot_iso_redfish() call for {import,export}_cluster()"`. I claim that "sometimes" holding the reference caused this bug (aside that the code is unused/untested). See also how dc9e029a78ff will violate your statement and there will be a BMC without a Host. The code just creates the BMC and uses it. Of course, we could instead create the BMC, attach it to the Host, and use the BMC via the Host. Then the statement would again hold. But why is that better?


IMO `Host` is only a handle to invoke commands on a host. As such, attaching BMC to it is wrong. Instead, if we want to bundle the BMC with other host-specific data, it should be on the `ClusterHost` or `ClusterNode`.  Note that in most cases, it is the `ClusterNode` object that creates the `BMC` and passes it on the the `Host`. `ClusterNode` should instead handle the `BMC` itself -- as happens on this pull request.

Reply:
We should then have something that holds both host and bmc. That makes more sense. When I did most of the refactoring to get BMC out of host, I intended to use a bmc from a REPL to control systems and when `ClusterHost` got introduced, it crossed my mind that we should move a lot of things from host there including bmc.

Reply:
> We should then have something that holds both host and bmc.

`ClusterHost` is already the main creator of `BMC` (and passes it to `Host`). It could easily keep the reference to the BMC for later. However, it is not really needed later.

Attaching it to the `ClusterHost` (or anywhere else) means to have additional state in an object. Instead, we can pass it directly to the few(!) places that will need it. If you look at the pull request, it's easy to see where BMC are created, and follow as they are passed around.

Reply:
You're saying that `ClusterHost` is the main creator of `BMC`, but I don't see it after this change. I don't see `ClusterHost` keeping an instance of `BMC`.